src/adapter/src/coord/timeline.rs

```diff
@@ -82,93 +79,59 @@ impl<T: CoordTimestamp> TimestampOracle<T> {
         F: Fn() -> T + 'static,
     {
         Self {
-            state: TimestampOracleState::Writing(initially.clone()),
-            advance_to: Some(initially),
+            state: TimestampOracleState {
+                read_ts: initially.clone(),
+                write_ts: initially,
+            },
             next: Box::new(next),
         }
     }
 
-    /// Peek the current value of the timestamp.
-    ///
-    /// No operations should be assigned to the timestamp returned by this function. The
-    /// timestamp returned should only be used to compare the progress of the `TimestampOracle`
-    /// against some external source of time.
-    ///
-    /// Subsequent values of `self.read_ts()` and `self.write_ts()` will be greater or equal to
-    /// this timestamp.
-    ///
-    /// NOTE: This can be removed if DDL is included in group commits.
-    fn peek_ts(&self) -> T {
-        match &self.state {
-            TimestampOracleState::Writing(ts) | TimestampOracleState::Reading(ts) => ts.clone(),
-        }
-    }
-
     /// Acquire a new timestamp for writing.
     ///
     /// This timestamp will be strictly greater than all prior values of
-    /// `self.read_ts()`, and less than or equal to all subsequent values of
     /// `self.read_ts()`.
-    fn write_ts(&mut self) -> T {
-        match &self.state {
-            TimestampOracleState::Writing(ts) => ts.clone(),
-            TimestampOracleState::Reading(ts) => {
-                let mut next = (self.next)();
-                if next.less_equal(&ts) {
-                    next = ts.step_forward();
-                }
-                assert!(ts.less_than(&next));
-                self.state = TimestampOracleState::Writing(next.clone());
-                self.advance_to = Some(next.clone());
-                next
-            }
+    fn write_ts(&mut self) -> WriteTimestamp<T> {
+        let mut next = (self.next)();
+        if next.less_equal(&self.state.write_ts) {
+            next = self.state.write_ts.step_forward();
         }
```

Comment:
I'm not sure I follow this. I think this was inherited from the `Reading(ts)` case above, but I think here it is not as necessary? Or if it is, the comparison should be with `self.state.read_ts` instead of `write_ts`, as we want to be sure the write occurs after any previously issued read timestamp

Reply:
Maybe this is an abstraction leak but my thought process was the following. You can't every call `append` twice with the same timestamp, so to prevent that from happening every write timestamp is larger than the previous write timestamp returned (this should be reflected in the doccomments, which it's not currently).